### PR TITLE
Error logging verbosity is not configurable for routing and used to determine log content of ErrorHandler (#16).

### PR DESCRIPTION
**Pull request by rbudzko**
Reused already existing property type. Set to full logging in tests as event filters asserts on that.

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

**Comment by rbudzko** [comment]
@ktoso, I see the mistake, but would like to understand why do you keep additional set of tests under doc/rst?

**Comment by ktoso** [comment]
> @ktoso, I see the mistake, but would like to understand why do you keep additional set of tests under doc/rst?

Because we're not done migrating things yet 😉

**Comment by rbudzko** [comment]
> Because we're not done migrating things yet :wink:
Sounds legit :shipit: 

I think it failed due to the timeout you mentioned last time, right?

**Comment by rbudzko** [comment]
PLS BUILD

**Comment by jrudolph** [comment]
PLS BUILD

**Comment by rbudzko** [comment]
> java.util.concurrent.TimeoutException: Futures timed out after [500 milliseconds]

@ktoso believes it is timing issue. I'm not sure this is related to this PR 😢. What do we do?

File: akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala [hunk 1]
```diff
@@ -37,16 +37,29 @@ object ExceptionHandler {
...
+
   def default(settings: RoutingSettings): ExceptionHandler =
     apply(knownToBeSealed = true) {
       case IllegalRequestException(info, status) ⇒ ctx ⇒ {
-        ctx.log.warning(
-          "Illegal request {}\n\t{}\n\tCompleting with '{}' response",
-          ctx.request, info.formatPretty, status)
+        log(
+          settings.errorLoggingVerbosity,
+          ctx.log.warning("Illegal request completing with '{}' response", status),
```
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala]: Why did you remove the newline?

File: akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala [hunk 2]
```diff
@@ -37,16 +37,29 @@ object ExceptionHandler {
...
+          ctx.log.warning("Illegal request {}\n\t{}\n\tCompleting with '{}' response", ctx.request, info.formatPretty, status))
+
         ctx.complete((status, info.format(settings.verboseErrorMessages)))
       }
       case NonFatal(e) ⇒ ctx ⇒ {
-        ctx.log.error(e, "Error during processing of request {}", ctx.request)
+        log(
+          settings.errorLoggingVerbosity,
+          ctx.log.error("Error during processing one of requests occurred due to '{}'", e.getMessage),
```
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala]:
```scala
ctx.log.error("Error during processing of request: '{}'", e.getMessage)
```
?

File: akka-http/src/main/resources/reference.conf [hunk 3]
```diff
@@ -13,6 +13,15 @@ akka.http.routing {
   # (Note that akka-http will always produce log messages containing the full error details)
   verbose-error-messages = off
 
+  # Configures the verbosity with which message (request or response) parsing
+  # errors are written to the application log.
+  #
+  # Supported settings:
+  # `off`   : no log messages are produced
+  # `simple`: a condensed single-line message is logged
+  # `full`  : the full error details (potentially spanning several lines) are logged
```
Comment (2beaucoup) [inline akka-http/src/main/resources/reference.conf]: s/potentially spanning several lines/including stack traces?

File: akka-http-core/src/main/scala/akka/http/impl/settings/RoutingSettingsImpl.scala [hunk 4]
```diff
@@ -28,6 +30,6 @@ object RoutingSettingsImpl extends SettingsCompanion[RoutingSettingsImpl]("akka.
     c getInt "range-count-limit",
     c getBytes "range-coalescing-threshold",
     c getIntBytes "decode-max-bytes-per-chunk",
-    c getString "file-io-dispatcher")
-
-}
+    c getString "file-io-dispatcher",
+    ErrorLoggingVerbosity(c getString "error-logging-verbosity"))
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/settings/RoutingSettingsImpl.scala]: Feels weird to reuse `ParserSettings.ErrorLoggingVerbosity` as this has no relation to parsing. Should we make `ErrorLoggingVerbosity` more general?

File: akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala (hunk 1 above)
Comment (rbudzko) [inline akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala]: Because this warning will be used for Simple which is meant to be one liner. At least it is how this is used in other places of akka-http.

File: akka-http/src/main/resources/reference.conf (hunk 3 above)
Comment (rbudzko) [inline akka-http/src/main/resources/reference.conf]: Yes. I'll upgrade this explanation in all reference.confs.

File: akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala (hunk 2 above)
Comment (rbudzko) [inline akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala]: Again, Simple is meant to be one liner. Are you unhappy with e.getMessage? How do we replace it for one liner?

File: akka-http-core/src/main/scala/akka/http/impl/settings/RoutingSettingsImpl.scala (hunk 4 above)
Comment (rbudzko) [inline akka-http-core/src/main/scala/akka/http/impl/settings/RoutingSettingsImpl.scala]: I like this name, but people with more stars 😉 should decide on that.

File: akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala (hunk 2 above)
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala]: just a change in wording

File: akka-http-core/src/main/scala/akka/http/impl/settings/RoutingSettingsImpl.scala (hunk 4 above)
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/settings/RoutingSettingsImpl.scala]: I too like the name. My suggestion was moving it to another place. ;)

Comment (rbudzko) [inline akka-http-core/src/main/scala/akka/http/impl/settings/RoutingSettingsImpl.scala]: Ah you mean `ParserSettings`. All right, I agree. It is worth noting RoutingSettings and ServerSettings do exist as well. Those two seems better candidates for more general `ErrorLoggingVerbosity`.

File: akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala (hunk 2 above)
Comment (rbudzko) [inline akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala]: It seems my eyes replace string content with same object SomeStringLiteral...

File: akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala (hunk 1 above)
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala]: I see: The previous version is now used for the `Full` case. I'd tweak this slightly to

     "Illegal request. Completing with {} response."

File: akka-http-core/src/main/scala/akka/http/impl/settings/RoutingSettingsImpl.scala (hunk 4 above)
Comment (rbudzko) [inline akka-http-core/src/main/scala/akka/http/impl/settings/RoutingSettingsImpl.scala]: I don't think we can remove `ErrorLoggingVerbosity` from `ParserSettings`, because of `Public API but not intended for subclassing`. We could copy it into `RoutingSettings`, but it is kind of meh. We also could create similar object in `RoutingSettings`, but we cannot remove or move current `ErrorLoggingVerbosity` due to `*Settings` being public.

**Comment by rbudzko** [comment]
`X-Real-Ip` failed. Passes locally 🐹 .

**Comment by 2beaucoup** [comment]
`Ok(X-Real-Ip: 1.2.3.4,List()) did not equal Ok(X-Real-Ip: 1.2.3.4,List()) (HttpHeaderSpec.scala:691)` :confused:

**Comment by rbudzko** [comment]
Sounds like Polish government claiming black is not black 👺 !

**Comment by rbudzko** [comment]
@2beaucoup @ktoso Can we re run this build? Should something still be improved?

File: akka-http-core/src/main/resources/reference.conf [hunk 5]
```diff
@@ -330,7 +330,7 @@ akka.http {
     # Supported settings:
     # `off`   : no log messages are produced
     # `simple`: a condensed single-line message is logged
-    # `full`  : the full error details (potentially spanning several lines) are logged
+    # `full`  : the full error details (potentially spanning several lines/including stack traces) are logged
```
Comment (2beaucoup) [inline akka-http-core/src/main/resources/reference.conf]: With my former comment I actually suggested replacing "potentially spanning several lines" with "including stack traces".

**Comment by rbudzko** [comment]
@2beaucoup Simplified it to just hiding request body 😢 .

File: akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala [hunk 6]
```diff
@@ -40,13 +40,11 @@ object ExceptionHandler {
...
       case IllegalRequestException(info, status) ⇒ ctx ⇒ {
-        ctx.log.warning(
-          "Illegal request {}\n\t{}\n\tCompleting with '{}' response",
-          ctx.request, info.formatPretty, status)
+        ctx.log.warning("Illegal request. Completing with {} response.", status)
         ctx.complete((status, info.format(settings.verboseErrorMessages)))
       }
       case NonFatal(e) ⇒ ctx ⇒ {
-        ctx.log.error(e, "Error during processing of request {}", ctx.request)
+        ctx.log.error(e, "Error during processing of request.", ctx.request)
```
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala]: `ctx.request` seems unused.  Either remove or change to something else and add back `{}`.

File: akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
  */
-
 package akka.http.scaladsl.server
 
 import scala.util.control.NonFatal
-import akka.http.scaladsl.settings.RoutingSettings
+import akka.http.scaladsl.settings.{ ParserSettings, RoutingSettings }
 import akka.http.scaladsl.model._
 import StatusCodes._
+import akka.http.scaladsl.settings.ParserSettings.ErrorLoggingVerbosity
```
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala]: All import changes can be reverted.

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/ExecutionDirectivesSpec.scala [hunk 8]
```diff
@@ -126,6 +126,32 @@ class ExecutionDirectivesSpec extends RoutingSpec {
     }
   }
 
+  "Default handler" should {
+    "handle `IllegalRequestException` with appropriate block of `ErrorHandler`" in EventFilter.warning(
+      occurrences = 1,
+      start = "Illegal request"
```
Comment (2beaucoup) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/ExecutionDirectivesSpec.scala]: use `message` instead of `start` for an exact match?

File: akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala (hunk 6 above)
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala]: or something like
```scala
ctx.log.error("Error during processing of request: '{}'. Completing with {} response.",
   e.getMessage, status)
```
as discussed earlier.

**Comment by rbudzko** [comment]
@2beaucoup Thnx for your points. I've applied requested changes. The build failed on some timeout.

File: akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala [hunk 9]
```diff
@@ -40,13 +39,11 @@ object ExceptionHandler {
   def default(settings: RoutingSettings): ExceptionHandler =
     apply(knownToBeSealed = true) {
       case IllegalRequestException(info, status) ⇒ ctx ⇒ {
-        ctx.log.warning(
-          "Illegal request {}\n\t{}\n\tCompleting with '{}' response",
-          ctx.request, info.formatPretty, status)
+        ctx.log.warning("Illegal request. Completing with {} response.", status)
```
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala]: What do you think of
```scala
ctx.log.warning("Illegal request: {}. Completing with {} response.", info.summary, status)
```
to streamline with processing errors?

**Comment by rbudzko** [comment]
@2beaucoup You are right. It's better now 😉 .

File: akka-http-core/src/main/resources/reference.conf [hunk 10]
```diff
@@ -330,7 +330,7 @@ akka.http {
     # Supported settings:
     # `off`   : no log messages are produced
     # `simple`: a condensed single-line message is logged
-    # `full`  : the full error details (potentially spanning several lines) are logged
+    # `full`  : the full error details (including stack traces) are logged
```
Comment (2beaucoup) [inline akka-http-core/src/main/resources/reference.conf]: I think this does no longer apply, now.

Comment (rbudzko) [inline akka-http-core/src/main/resources/reference.conf]: Good question. It's about parsing. I'll check.

Comment (ktoso) [inline akka-http-core/src/main/resources/reference.conf]: Ok, this seems to have been addressed

**Comment by ktoso** [comment]
LGTM, thanks! Seems all of @2beaucoup's comments were addressed as well.

**Comment by rbudzko** [comment]
Yay, I think I'll open some bottle after this one 😉 .

**Comment by 2beaucoup** [comment]
Well deserved! Cheers @rbudzko! 🍻

**Comment by rbudzko** [comment]
Refs #16.
